### PR TITLE
test: add E2E integration tests with real embedder

### LINKED ISSUE #56: test: 統合テストの追加
## 概要

`tests/` ディレクトリに Rust の統合テスト（`#[test]` in `tests/*.rs`）を追加し、エンドツーエンドの動作を検証する。

## 背景・動機

- 現在 405 のユニットテストがあるが、すべてモジュール内の `#[cfg(test)]` で定義されている
- `tests/` ディレクトリに `.rs` ファイルが1つもなく、モジュール横断の統合テストが存在しない
- `tests/search-test-patterns.md` にテストシナリオが文書化されているが、自動化されていない
- CI は `cargo test` + `cargo llvm-cov --fail-under-lines 90` を実行しており、`tests/*.rs` を追加すれば自動的に CI で実行される

## 課題

- index → search → 結果検証のエンドツーエンドフローが自動テストされていない
- daemon protocol（ソケット経由のリクエスト/レスポンス）の結合テストがない（`daemon.rs` のテストは `handle_request` を直接呼んでおり、ソケット層を経由していない）
- watcher スレッドの実際のファイル変更検知 → インデックス更新の流れがテストされていない
- `search-test-patterns.md` の E1-E5, T1-T7, F1-F4, EC1-EC5 が手動検証のまま

### テストインフラ

- `tests/` ディレクトリに統合テストファイルを配置
- `test_utils::setup_db_with_dir()` + `db::get_memory_connection()` を活用してインメモリ DB + tempdir で実行
- embedder 依存のテスト（ベクトル検索）は embedder ソケットがない環境でスキップ（`#[ignore]` または feature gate）
- 全テストがローカル `cargo test` と CI の両方で再現可能であること

### テストカテゴリ（優先順）

1. **index → search ラウンドトリップ** — .md ファイルをインデックスして FTS5 検索で取得できることを検証
2. **daemon protocol ラウンドトリップ** — UNIX ソケット経由で Index/Search/Status/Doctor リクエストを送受信
3. **temporal search** — `search-test-patterns.md` の T1-T7 を自動化
4. **entity search** — E1-E5 を自動化
5. **edge cases** — EC1-EC5 を自動化
6. **watcher スレッド** — tempdir にファイルを作成/変更/削除し、DB に反映されることを検証

### ファイル構成案

```text
tests/
├── search-test-patterns.md   # 既存ドキュメント
├── index_search.rs           # index → search ラウンドトリップ
├── daemon_protocol.rs        # ソケット経由のリクエスト/レスポンス
├── temporal.rs               # 時間フィルタ検索
└── common/
    └── mod.rs                # 共通ヘルパー（テスト用 .md ファイル生成等）
```

## 影響範囲

- 既存コードへの変更なし（テスト追加のみ）
- CI の `cargo test` で自動実行される
- `cargo llvm-cov` のカバレッジ計測に含まれる

## DoD (Definition of Done)

- [ ] `tests/` に統合テストファイルが追加されている
- [ ] index → search のラウンドトリップテストが動作する
- [ ] daemon protocol のソケットラウンドトリップテストが動作する
- [ ] `search-test-patterns.md` のシナリオが少なくとも F1-F4, EC1-EC5 で自動化されている
- [ ] embedder 非依存でローカル実行可能
- [ ] CI がパスしている

### PR DESCRIPTION
**Pull request by key**
## Summary

- Shell-based E2E tests that exercise the full CLI surface (`tsm`/`tsmd`) without modifying application code
- Uses public domain texts from Aozora Bunko (夏目漱石, 宮沢賢治, 太宰治) as test data
- Separate GHA workflow (`e2e.yml`) with model caching for ruri-v3-30m
- Tests run with real embedder for true end-to-end verification

### Test categories
- Daemon basics (status, doctor)
- Index → Search round-trip
- FTS5 full-text search
- Entity search (tag boost)
- Temporal search (`--recent`, `--year`, `--after`/`--before`)
- Vector search (semantic similarity via ruri-v3-30m)
- Dictionary (add word → rebuild FTS → verify search)
- Edge cases (empty query, single char, invalid args)
- File watcher (add/delete detection)

## Test plan
- [ ] E2E workflow triggers on this PR
- [ ] All test cases pass in GHA
- [ ] Model caching works (second run should skip download)

Closes #56

### CODE REVIEW

File: tests/e2e.sh [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+echo ""
+log "=== Daemon basics ==="
+
+OUTPUT=$(tsm status 2>&1) || true
+EXIT=$?
+assert_contains "status: daemon running" "Daemon:" "$OUTPUT" "$EXIT"
+
+OUTPUT=$(tsm doctor -f json 2>&1) || true
+EXIT=$?
+assert_json "doctor: json output" '.issue_count >= 0' "$OUTPUT" "$EXIT"
```
[the remark this answers is not in the document]
Comment (key) [reply tests/e2e.sh]: Fixed in 6226b5a.

File: tests/e2e.sh [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+log "=== Dictionary ==="
+
+# Dict test: add a word to user dict, rebuild FTS, verify search works with it.
+# Use "セリヌンティウス" — a proper noun only in hashire-melos.md that lindera
+# splits into multiple tokens by default, but as a dict entry becomes one token.
+DICT_WORD="セリヌンティウス"
+DICT_FILE="hashire-melos"
+
+# Search before dict registration — should already hit via constituent tokens
+OUTPUT_BEFORE=$(search_json "$DICT_WORD" --fallback fts-only 2>/dev/null) || true
```
[the remark this answers is not in the document]
Comment (key) [reply tests/e2e.sh]: Not applicable — the CLI was updated to accept kebab-case `fts-only` in e2f3d45. Tests pass with the current syntax.

File: tests/e2e.sh [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+
+# Wait for daemon ready
+sleep 3
+
+# Verify search still works after the stop→rebuild→start cycle
+OUTPUT_POST=$(search_json "メロス 激怒" --fallback fts-only) || true
+EXIT=$?
+assert_json "dict: search works after rebuild" \
+    'any(.[]; .source_file | contains("hashire-melos"))' "$OUTPUT_POST" "$EXIT"
+
```
[the remark this answers is not in the document]
Comment (key) [reply tests/e2e.sh:332]: Same as above — `fts-only` (kebab-case) is the correct syntax after e2f3d45.

File: tests/e2e.sh [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+# Dict test: add a word to user dict, rebuild FTS, verify search works with it.
+# Use "セリヌンティウス" — a proper noun only in hashire-melos.md that lindera
+# splits into multiple tokens by default, but as a dict entry becomes one token.
+DICT_WORD="セリヌンティウス"
+DICT_FILE="hashire-melos"
+
+# Search before dict registration — should already hit via constituent tokens
+OUTPUT_BEFORE=$(search_json "$DICT_WORD" --fallback fts-only 2>/dev/null) || true
+
```
[the remark this answers is not in the document]
Comment (key) [reply tests/e2e.sh:305]: Fixed in 0b8e44e. Added `assert_json` on `OUTPUT_BEFORE` to verify pre-registration search found results.

File: tests/e2e.sh [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+else
+    # Empty query might be rejected by clap, that's OK too
+    pass "edge: empty query → handled (exit $EXIT)"
+fi
+
+# EC4: Single character
+OUTPUT=$(search_json "a" 2>/dev/null) || true
+EXIT=$?
+# Should not crash; any exit code is fine as long as it doesn't segfault
+pass "edge: single char 'a' → no crash (exit $EXIT)"
```
[the remark this answers is not in the document]
Comment (key) [reply tests/e2e.sh]: Fixed in 0b8e44e. Now uses `assert_json` to verify exit 0 and valid JSON array.

File: tests/e2e.sh [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+
+# Compute dynamic dates
+TODAY=$(date +%Y-%m-%d)
+ONE_YEAR_AGO=$(date -d '1 year ago' +%Y-%m-%d)
+THREE_MONTHS_AGO=$(date -d '3 months ago' +%Y-%m-%d)
+LAST_YEAR=$(date -d '1 year ago' +%Y)
+THREE_MONTHS_AGO_START=$(date -d '3 months ago' +%Y-%m-01)
+# End of that month: first day of (3 months ago + 1 month)
+THREE_MONTHS_AGO_END=$(date -d '2 months ago' +%Y-%m-01)
+
```
[the remark this answers is not in the document]
Comment (key) [reply tests/e2e.sh:129]: Fixed in 6226b5a. Added 'Linux with GNU coreutils' to prerequisites. This script runs only on GHA ubuntu-latest.

File: tests/e2e.sh [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+export TSM_EMBEDDER_IDLE_TIMEOUT=0
+export TSM_EMBEDDER_BACKFILL_INTERVAL=0
+
+# Compute dynamic dates
+TODAY=$(date +%Y-%m-%d)
+ONE_YEAR_AGO=$(date -d '1 year ago' +%Y-%m-%d)
+THREE_MONTHS_AGO=$(date -d '3 months ago' +%Y-%m-%d)
+LAST_YEAR=$(date -d '1 year ago' +%Y)
+THREE_MONTHS_AGO_START=$(date -d '3 months ago' +%Y-%m-01)
+# End of that month: first day of (3 months ago + 1 month)
```
[the remark this answers is not in the document]
Comment (key) [reply tests/e2e.sh]: Fixed in 6226b5a. Comment updated to 'Exclusive upper bound: first day of (3 months ago + 1 month)'.

File: tests/e2e.sh [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+export TSM_INDEX_ROOT
+TSM_INDEX_ROOT="$(mktemp -d)"
+export TSM_EMBEDDER_IDLE_TIMEOUT=0
+export TSM_EMBEDDER_BACKFILL_INTERVAL=0
+
+# Compute dynamic dates
+TODAY=$(date +%Y-%m-%d)
+ONE_YEAR_AGO=$(date -d '1 year ago' +%Y-%m-%d)
+THREE_MONTHS_AGO=$(date -d '3 months ago' +%Y-%m-%d)
+LAST_YEAR=$(date -d '1 year ago' +%Y)
```
[the remark this answers is not in the document]
Comment (key) [reply tests/e2e.sh]: Fixed in 6226b5a. Removed.

File: tests/e2e.sh [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+# ── Dictionary test ───────────────────────────────────────────────────
+
+echo ""
+log "=== Dictionary ==="
+
+# Dict test: add a word to user dict, rebuild FTS, verify search works with it.
+# Use "セリヌンティウス" — a proper noun only in hashire-melos.md that lindera
+# splits into multiple tokens by default, but as a dict entry becomes one token.
+DICT_WORD="セリヌンティウス"
+DICT_FILE="hashire-melos"
```
[the remark this answers is not in the document]
Comment (key) [reply tests/e2e.sh]: Fixed in 6226b5a. Removed.
